internal/pipeline/executor.go

```diff
@@ -36,161 +35,90 @@ type Executor struct {
-	if p.pipeline == nil {
-		// READ the pipeline in the pkg.
-		p.pipeline = &kptfilev1alpha2.Pipeline{}
-	}
-	return p.pipeline
-}
-
-// resolveSources takes a list of sources (./*, ./, ...) and current pkg path
-// and returns list of package paths that this pkg depends on.
-// This is one of the critical pieces of code
-func (p *pkg) resolveSources() ([]string, error) {
-	var pkgPaths []string
-	paths, err := resolveSource(sourceAllSubPkgs, p.Path())
-	if err != nil {
-		return nil, err
+// newPkgNode returns a pkgNode instance given a path or pkg.
+func newPkgNode(path string, p *pkg.Pkg) (pn *pkgNode, err error) {
+	if path == "" && p == nil {
+		return pn, fmt.Errorf("missing package path %s or package", path)
 	}
-	pkgPaths = append(pkgPaths, paths...)
-	return pkgPaths, nil
-}
-
-func resolveSource(source string, pkgPath string) ([]string, error) {
-	switch source {
-	case sourceCurrentPkg:
-		// include only this pkg sources
-		return []string{pkgPath}, nil
-	case sourceAllSubPkgs:
-		// including current pkg and subpkgs in current directory
-		var paths []string
-		paths = append(paths, pkgPath)
-		files, err := ioutil.ReadDir(pkgPath)
+	if path != "" {
+		p, err = pkg.New(path)
 		if err != nil {
-			return nil, err
-		}
-		for _, f := range files {
-			if f.IsDir() {
-				// A directory is a package if it has a Kptfile.
-				// This may change as the concept of a package is expanded such that
-				// every directory is its own kpt package
-				absolutePath := path.Join(pkgPath, f.Name())
-				subPaths, err := pathutil.DirsWithFile(absolutePath, kptfilev1alpha2.KptFileName, false)
-				if err != nil {
-					return nil, err
-				}
-				sort.Strings(subPaths)
-				paths = append(paths, subPaths...)
-			}
+			return pn, fmt.Errorf("failed to read package %w", err)
 		}
-		return paths, nil
-	default:
-		// s points to a specific sub pkg
-		return []string{source}, nil
 	}
-}
-
-// localResources reads
-func (p *pkg) localResources(includeMetadata bool) (resources []*yaml.RNode, err error) {
-	// TODO(droot): figure out how to avoid reading sub packages here ?
-	pkgReader := &kio.LocalPackageReader{PackagePath: p.Path(), MatchFilesGlob: kio.MatchAll}
-	resources, err = pkgReader.Read()
-	if err != nil {
-		err = fmt.Errorf("failed to read resources for pkg %s %w", p.Path(), err)
-		return resources, err
+	// Note: Ensuring the presence of Kptfile can probably be moved
+	// to the lower level pkg abstraction, but not sure if that
+	// is desired in all the cases. So revisit this.
+	if _, err = p.Kptfile(); err != nil {
+		return pn, fmt.Errorf("failed to read kptfile for package %s %w", p, err)
 	}
-	if !includeMetadata {
-		// TODO(droot): this will be the place where we filter kpt resources
-		resources = filterMetaData(resources)
+	pn = &pkgNode{
+		pkg:   p,
+		state: Dry, // package starts in dry state
 	}
-	return resources, err
+	return pn, nil
+}
+
+func (pn *pkgNode) Path() string {
+	return string(pn.pkg.UniquePath)
```

Comment:
I agree. Last time I tried using `UniquePath`, I was running into issues where I had to convert `UniquePath` to string in ton of places because hydration use filesystem package extensively, so I gave up. Now that I have code almost final, I want to try again. I am keeping this comment open.

Reply:
Made another attempt today to use `pkg.UniquePath` type and worked pretty well this time. Tracked these changes in a separate commit to make easier to review.